src/v/iceberg/tests/metadata_query_test.cc

```diff
@@ -0,0 +1,419 @@
+        auto r_mf = executor.execute_query(q_mf).get();
+        auto r_m = executor.execute_query(q_m).get();
+
+        return std::make_tuple(
+          std::move(r_s.assume_value()),
+          std::move(r_mf.assume_value()),
+          std::move(r_m.assume_value()));
+    }
+
+    std::unique_ptr<cloud_io::scoped_remote> sr;
+    manifest_io io;
+};
+
+using namespace testing;
+TEST_F(MetadataQueryTest, TestEmptyQueryMatchesEverything) {
+    auto tx = generate_metadata().get();
+    auto [snapshots, files, manifests] = execute_for_all_results(tx.table());
+    ASSERT_EQ(snapshots.size(), 5);
+    // all snapshots should be present in the result
+    ASSERT_EQ(snapshots, tx.table().snapshots);
+
+    // all manifest_files should be present in result
+    ASSERT_EQ(files, collect_all_manifest_files(tx.table()).get());
+
+    // all manifests should be present in result
+    auto all = collect_all_manifests(tx.table()).get();
+    ASSERT_EQ(manifests, collect_all_manifests(tx.table()).get());
+}
+
+bool is_even(int v) { return v % 2 == 0; }
+
+TEST_F(MetadataQueryTest, TestQuerySnapshots) {
+    auto tx = generate_metadata().get();
+    // snapshot query, only those snapshots which has even sequence number,
+    // those will be snapshot 2 and 4
+    auto s_matcher = [](const snapshot& s) {
+        return s.sequence_number() % 2 == 0;
+    };
+
+    auto [snapshots, files, manifests] = execute_for_all_results(
+      tx.table(), s_matcher);
+    ASSERT_EQ(snapshots.size(), 2);
+    EXPECT_THAT(
+      snapshots, Each(Field(&snapshot::sequence_number, Truly(is_even))));
+
+    auto all_manifest_files = collect_all_manifest_files(tx.table()).get();
+
+    /**
+     * The ordering here is counter intuitive.
+     * Table metadata contains 5 snapshots [s1,s2,s3,s4,s5] each snapshot
+     * contains metadata files
+     *
+     * s1 - [f0]
+     * s2 - [f1,f0]
+     * s3 - [f2,f1,f0]
+     * s4 - [f3,f2,f1,f0]
+     * s5 - [f4,f3,f2,f1,f0]
+     *
+     * When test util function collects all files/manifests it traverses the
+     * snapshot list and deduplicate the files. The result of this operation is:
+     * [f0,f1,f2,f3,f4]
+     *
+     * When query is executed with is even segment sequence it matches s2 and s4
+     * from s2 it takes
+     *
+     * [f1, f0] and then from s4 [f3,f2]
+     *
+     * So the end result is [f1,f0,f3,f2].
+     */
+    ASSERT_EQ(files.size(), 4);
+    ASSERT_EQ(files[0], all_manifest_files[1]);
+    ASSERT_EQ(files[1], all_manifest_files[0]);
+    ASSERT_EQ(files[2], all_manifest_files[3]);
+    ASSERT_EQ(files[3], all_manifest_files[2]);
+
+    // the same ordering rule applies to the manifests
+    auto all_manifests = collect_all_manifests(tx.table()).get();
+    ASSERT_EQ(manifests.size(), 4);
+    ASSERT_EQ(manifests[0], all_manifests[1]);
+    ASSERT_EQ(manifests[1], all_manifests[0]);
+    ASSERT_EQ(manifests[2], all_manifests[3]);
+    ASSERT_EQ(manifests[3], all_manifests[2]);
```

Comment:
Same as above, use `ASSERT_THAT` and `ElementsAre`. `ElementsAre` requires proper ordering.